maint: add maint/gen_cross.pl to help generating crossfile

hzhou opened the pull request:
## Pull Request Description
Provide a script that a user can run to generate cross_values.txt that
can be passed to MPICH configure with --with-cross=path/to/cross_values.txt.

Fixes #3753


## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
```
$ perl maint/gen_cross.pl  -h
  Run this script on the host system to generate a cross file for
  MPICH configure. Set compiler via CC and FC environment variables.
  You may also use the generated file as template and manually set
  the type sizes.

$ CC=clang FC=gfortran perl maint/gen_cross.pl
CC: clang
FC: gfortran
  --> [cross_values.txt]
ac_cv_sizeof_void_p=8
ac_cv_sizeof_char=1
ac_cv_sizeof_short=2
ac_cv_sizeof_int=4
ac_cv_sizeof_long=8
ac_cv_sizeof_long_long=8
ac_cv_sizeof_size_t=8
ac_cv_sizeof_off_t=8
ac_cv_sizeof_float=4
ac_cv_sizeof_double=8
ac_cv_sizeof_long_double=16
CROSS_F77_SIZEOF_INTEGER=4
CROSS_F77_SIZEOF_REAL=4
CROSS_F77_SIZEOF_DOUBLE_PRECISION=8
CROSS_F77_TRUE_VALUE=1
CROSS_F77_FALSE_VALUE=0
CROSS_F90_ADDRESS_KIND=8
CROSS_F90_OFFSET_KIND=8
CROSS_F90_INTEGER_KIND=4
CROSS_F90_REAL_MODEL=6,37
CROSS_F90_DOUBLE_MODEL=15,307
CROSS_F90_INTEGER_MODEL=9
CROSS_F90_ALL_INTEGER_MODELS="2, 1, 4, 2, 9, 4, 18, 8"
CROSS_F90_INTEGER_MODEL_MAP="{ 2 , 1 , 1 }, { 4 , 2 , 2 }, { 9 , 4 , 4 }, { 18 , 8 , 8 }"

Crossfile generated in cross_values.txt.
Pass to MPICH configure with --with-cross=path/to/cross_values.txt.
```

raffenet commented on configure.ac:
```diff
@@ -2896,11 +2896,12 @@ for type in short int long float double \
-system; this program outputs an appropriate file for the --with-cross option])
+use the --with-cross=file configure option to specify a file containing
+Bourne (sh) shell assignments to sizeof values for all datatype types.
+The Perl script maint/gen_cross.pl can be compiled and run on the target
```
```suggestion
The Perl script maint/gen_cross.pl can be run on the target
```

raffenet reviewed (commented):
I would complain about adding new Perl scripts instead of Python, but we're pretty outside the normal development path here so I think its fine. I did run this on my laptop and got a bunch of compile errors. Are these expected?
```
CC: gcc
FC: gfortran
t.f:12:30:

   11 |        print *, 'A0:', cisize(A0(1), A0(2))
      |                              2
   12 |        print *, 'A1:', cisize(A1(1), A1(2))
      |                              1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(4)/INTEGER(4)).
t.f:13:30:

   11 |        print *, 'A0:', cisize(A0(1), A0(2))
      |                              2
   12 |        print *, 'A1:', cisize(A1(1), A1(2))
   13 |        print *, 'A2:', cisize(A2(1), A2(2))
      |                              1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(8)/INTEGER(4)).
t.f:14:30:

   11 |        print *, 'A0:', cisize(A0(1), A0(2))
      |                              2
......
   14 |        print *, 'A3:', cisize(A3(1), A3(2))
      |                              1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(1)/INTEGER(4)).
t.f:15:30:

   11 |        print *, 'A0:', cisize(A0(1), A0(2))
      |                              2
......
   15 |        print *, 'A4:', cisize(A4(1), A4(2))
      |                              1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(2)/INTEGER(4)).
t.f:17:30:

   11 |        print *, 'A0:', cisize(A0(1), A0(2))
      |                              2
......
   17 |        print *, 'A6:', cisize(A6(1), A6(2))
      |                              1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(8)/INTEGER(4)).
t.f90:8:25:

    8 | PRINT *, 'INTEGER_KIND:', SELECTED_INT_KIND()
      |                         1
Error: Missing actual argument 'r' in call to 'selected_int_kind' at (1)
```

hzhou commented:
> I would complain about adding new Perl scripts instead of Python

:) Python will take more effort (for me at least) than Perl, plus the usual python 3 portability issue. That said, if there is anyone who formally complains, I'll happily rewrite it into Python.

> Are these expected?

No. I will check and fix them.

hzhou commented:
The errors were complaining of missing `-fallow-argument-mismatch`.  I added additional help text to mention it. I also added options to choose different method of checking Fortran type sizes -- hybrid, direct, asm. Default to direct to avoid argument checking issue.
